src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivityTagsCollection.cs

```diff
@@ -59,12 +59,15 @@ public ActivityTagsCollection(IEnumerable<KeyValuePair<string, object>> list)
         ///     - Otherwise, a new item will get added to the collection.
         /// </summary>
         /// <value>Object mapped to the key</value>
-        public object this[string key]
+        [DisallowNull]
```

Comment:
>DisallowNull [](start = 9, length = 12)

I am now wondering what DisallowNull will do? does it will prevent the setter from using null values? of so then this will be wrong.

Reply:
> does it will prevent the setter from using null values

Yes, it'll warn if the value may be null. Nulls are allowed to be added as values? Then shouldn't Add take an `object?` rather than `object`? And implement the interface with a nullable TKey? I'm wondering why this setter is special

Reply:
the indexer can allow null values. null values for this collection means remove the entry matching the key. This is the behavior that OpenTelemetry suggested for such collection. only keys cannot be null. So we need to remove `DisallowNull` I guess.

Reply:
@buyaa-n reminded me in other comment the indexer can return null too. anyway, I think it make sense having the indexer is nullable and not using DisallowNull.

Reply:
What's the behavior when you call `Add(key, null);`?

Reply:
It will add the (key, null) to the collection. this is different than the index behavior. Add always add and used only if anyone want to have null value.

Reply:
Then the interface implementation is annotated incorrectly. (And the difference between the indexer and Add behavior is very confusing.)

Reply:
We discussed that in details in the design review and that what we concluded there. I agree we need to fix the interface annotation and I think that is what we are doing now.

Reply:
That is what i meant in my original comment https://github.com/dotnet/runtime/pull/41731#discussion_r482682472, making interface implementation allow nullable value need change for the entire dictionary. Which i was not sure if we want to do with this PR, if we want to udpate that with this PR i can do that

Reply:
> I agree we need to fix the interface annotation and I think that is what we are doing now.

I mean the whole interface, not just the indexer. Anywhere TValue is `object` needs to be changed to `object?`, including in the definition.

> We discussed that in details in the design review and that what we concluded there

I think that's not a good outcome. The only difference between setting a value into a dictionary and adding that value into the dictionary should be overwrite behavior.

Reply:
> I think that's not a good outcome. The only difference between setting a value into a dictionary and adding that value into the dictionary should be overwrite behavior.

There is more details around that because of all interfaces we had in Activity objects which allow adding tags with null values.

> Which i was not sure if we want to do with this PR, if we want to udpate that with this PR i can do that

I think we can do it in this PR. now need a new PR for that. I am not expecting the changes will be big anyway.